lib/ProductOpener/APIProductServices.pm

```diff
@@ -0,0 +1,180 @@
+# This file is part of Product Opener.
+#
+# Product Opener
+# Copyright (C) 2011-2023 Association Open Food Facts
+# Contact: contact@openfoodfacts.org
+# Address: 21 rue des Iles, 94100 Saint-Maur des Fossés, France
+#
+# Product Opener is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+=head1 NAME
+
+ProductOpener::APIProductServices - Microservices to enrich a product object
+
+=head1 DESCRIPTION
+
+This module implements a microservice API for operations done on an product object.
+
+Applications can send product data (for instance a nested list of ingredients),
+ask for one or more services to be executed on the input product data
+(for instance computing the min, max and estimated percentages of each ingredient),
+and get back resulting product data (possibly filtered to get only specific fields back).
+
+=head2 INTERFACE
+
+=head3 Request
+
+The Routing.pm and API.pm module offer an HTTP interface of this form:
+POST /api/v3/product_services/[comma separated list of services]
```

Comment:
Just a personal observation: Commas in the path to designate separate services doesn't feel that natural *to me*. A lot of frameworks allow multiple values for a single parameter, so that something like `/api/v3/product_services?service=echo&service=estimate_ingredients_percent` could work.

Reply:
I'd prefer not to have both query parameters + parameters in the JSON body.
Maybe we could just make the endpoint /api/v3/product_services and have a services array inside the JSON body instead.

Reply:
I'm personally not sure about having this very generic "product_services" name. It feels like each operation should be its own service, e.g. /api/v3/estimate_ingredients, /api/v3/calculate_nutriscore. If the caller wants to chain multiple processes then they can simply make multiple API calls. I think this will make it easier in the futures to switch the routes for specific services to go somewhere other than product opener.

Each service could take a product object or product id as input and then just return the specific fields that it computes.  I'm not sure whether the service itself should then update the product with the modified fields or whether that should be the caller's responsibility. In a "pure" microservice model each service would store just the information that it owns in its own logically separate database and then emit an event to let other services know that this data is now available. In situations where you want to be able to provide the entire set of information from a single API you could either replicate this data to a "consolidated product" service or have some form of API federation to call out to the different services and bring all of the data together (I think the former would be more resilient)

Reply:
I agree with hangy that naming the services in the query URL this way feels weird, I would much rather see them in the post body.
I also agree with john about the "product_services" name, although I would like to have all product-computation related operations under a single API path.
Why not using `/api/v3/product/*`? We could have:
- `/api/v3/product/estimate_ingredients` and other for each individual processing items
- `/api/v3/product/pipeline` if you want to chain processing items as you did here

Reply:
In my mind, the /api/v3/product_services is the federation API. You can use it to execute 1 or more services, either by specifying them individually, or by using aliases that bundle a bunch of operations together (like a Makefile). This way we can cut services in smaller nested services easily. e.g. we could start with one big "ingredients_analysis" service, that we could further cut in smaller services like ingredients_parsing, ingredients_percent_estimation, ingredients_vegan_vegetarian_analysis etc. (if we want to access only a specific part externally, or if internally we want to do something outside of product opener), without having to change all the callers: they can still call the "parent" service and it will be transparent to them.

It's important to have this federated pipeline, because that's how we will use those services, at least initially. e.g. for the Rate my recipe project, I need at least 5 services: ingredient parsing, ingredient percent estimation, nutrient estimation, nutriscore, ecoscore. It doesn't make a lot of sense to have the client make 5 back and forth requests, one after the other. And in practice the federated product_services API will know that all those services are provided internally by Product Opener, so it will use a fast path and not actually issue HTTP queries, those will be only needed for services provided by something else than Product Opener.

Reply:
Hi @stephanegigandet , I agree with your points. I think there are potentially two different threads here that have different requirements:

1. Provide services that other applications / services can consume to perform specific tasks. In these cases the calling application should not need to know what steps are involved in getting the data they need, they just want the result (synchronously). They potentially aren't updating a product in the database
2. In our own pipeline where we are saving a product and then chaining all of the various computations that are possible with the data that has been provided. In this case we could potentially adopt an asynchronous model to give better scalability, so each individual computation would be listening for the event that would enable them to complete their part in the chain and then emit an event when their computation is complete

You are mainly looking to address item 1 and so the approach you are taking makes sense for this. In which case I think my main comment would be that rather than the calling application have to know which steps need to be run it should just indicate the various states that it wants the results to be in, e.g. "nutriscore_computed", "ingredients_estimated", "ecoscore_computed". It would then be up to the federation API to work out which internal (or external) services need to be called, and in what order, to satisfy those states. That way the calling application does not need to be updated if we enhance our algorithms to require further inputs.

For example, if an application just needs the ecoscore_computed, then at the moment we just need a product category. However, in the future, with better ingredient estimation, we might enhance the ecoscore accuracy by looking at the specific ingredients, hence the ingredients estimation service would also need to be called. The calling application should not need to care about this change, it just wants us to return the ecoscore.

By the way, in this model it makes sense to me now that we potentially return a whole product object, or just the fields that the caller wants, although I still feel that internally our services should only return /modify the fields that they "own".

Reply:
> * `/api/v3/product/estimate_ingredients` and other for each individual processing items
> * `/api/v3/product/pipeline` if you want to chain processing items as you did here

So for now, I put the specification of services in the JSON of the request body, with a "services" array. So it's like the "pipeline" option above. If you just want one service, you just put only one service in the array.